### DISCUSSION TITLE
[Question] Trigger broadcast

### DISCUSSION THREAD

**Discussion by zabulus** [quartznet/quartznet#2598] [Q&A]
**Is your feature request related to a problem? Please describe.**
We're using Quartz in clustered environment and we would like to use it in broadcast scenario. Job triggered on one node is executed on all other nodes. In opposite to current behavior where job is executed on one node.

**Describe the solution you'd like**
**Describe alternatives you've considered**
I see here several alternatives:
1. Trigger pinning. During triggering explicitly specify node where job will execute. We have a membership mechanism, so it's possible just specify list of nodes to trigger job at.
2. Explicit broadcast. During trigger in cluster mode somehow specify that this should be executed on all nodes and forward responsibility to Quartz to figure everything itself.
3. Another option comes to my mind is distributed ISchedulerListener. Currently only listeners on the same node receive information about scheduling of a job. If other nodes could receive events about a job was triggered, they might trigger job on theirs side as well.

Currently we'll implement custom notification mechanism over tcp channel to trigger jobs, but this is something we would like to avoid.

**Comment by lahma** [comment]
Sounds like a cool idea, but first thing that comes to my mind is the problem of knowing what are "all nodes". Nodes are dynamic and during broadcast there might be new appearing and some disappearing, of course that would be quite highly unlike case but broadcast sounds like something that might happen often. 

Would a messaging solution with publish-subscribe work better in this case?

**Reply by zabulus** [reply]
> Would a messaging solution with publish-subscribe work better in this case?

It definitely would, but in our case that would be overkill. The described case indeed better aligns with responsibilities of message queues. That's why I proposed other solutions that look less like MQ but more specific for Quartz. 

> But first thing that comes to my mind is the problem of knowing what are "all nodes"

We already manage cluster membership by ourselves. If you think that would be problem for Quartz, you just declare that it is client's responsibility to specify nodes.

**Comment by zabulus** [comment]
I've managed to accomplish targeted execution using following code:
```
public class TargetAwareJobStore : JobStoreTX
{
    public string CurrentNodeId { get; set; }

    protected override async Task<IReadOnlyCollection<IOperableTrigger>> AcquireNextTrigger(
        ConnectionAndTransactionHolder conn,
        DateTimeOffset noLaterThan,
        int maxCount,
        TimeSpan timeWindow,
        CancellationToken cancellationToken = default)
    {
        var triggers = await base.AcquireNextTrigger(conn, noLaterThan, maxCount, timeWindow, cancellationToken).ConfigureAwait(false);

        if (triggers.Count == 0)
            return triggers;

        Log.Debug($"Filtering {triggers.Count} triggers for node {CurrentNodeId}");

        var triggerToRelease = new List<IOperableTrigger>();
        var resultTriggers = new List<IOperableTrigger>();

        foreach (var trigger in triggers)
        {
            var targetNode = trigger.JobDataMap.GetString(JobConstants.TargetNodeKey);
            var shouldExecute = string.IsNullOrEmpty(targetNode) || targetNode == CurrentNodeId;

            if (shouldExecute)
            {
                resultTriggers.Add(trigger);
            }
            else
            {
                Log.Debug($"Skipping trigger {trigger.Key} targeted for node {targetNode}");
                triggerToRelease.Add(trigger);
            }
        }

        if (triggerToRelease.Any())
        {
            foreach (var trigger in triggerToRelease)
            {
                await ReleaseAcquiredTrigger(conn, trigger, cancellationToken).ConfigureAwait(false);
            }
        }

        Log.Debug($"Filtered to {resultTriggers.Count} triggers for node {CurrentNodeId}");

        return resultTriggers;
    }
}
```
While this works in testing, I'm concerned about potential edge cases or issues that might not be immediately obvious. Please review this approach and point out any caveats or problems it might have.

**Reply by zabulus** [reply]
@lahma Hi there! Just following up on this discussion about implementing trigger broadcasting in Quartz. I've shared my implementation using a custom TargetAwareJobStore that filters triggers based on node targeting. Would appreciate your thoughts on this approach and any potential issues you might foresee with this implementation. Thanks for your time!
